src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -0,0 +1,436 @@
+   * @return a TiDAGRequest
+   */
+  public TiDAGRequest setStartTs(long startTs) {
+    this.startTs = startTs;
+    return this;
+  }
+
+  long getStartTs() {
+    return startTs;
+  }
+
+  /**
+   * set having clause to select query
+   *
+   * @param having is a expression represents Having
+   * @return a TiDAGRequest
+   */
+  public TiDAGRequest setHaving(TiExpr having) {
+    this.having = requireNonNull(having, "having is null");
+    return this;
+  }
+
+  public TiDAGRequest setDistinct(boolean distinct) {
+    this.distinct = distinct;
+    return this;
+  }
+
+  public boolean isDistinct() {
+    return distinct;
+  }
+
+  /**
+   * add aggregate function to select query
+   *
+   * @param expr is a TiUnaryFunction expression.
+   * @return a SelectBuilder
+   */
+  public TiDAGRequest addAggregate(TiExpr expr) {
+    requireNonNull(expr, "aggregation expr is null");
+    aggregates.add(Pair.create(expr, expr.getType()));
+    return this;
+  }
+
+  public TiDAGRequest addAggregate(TiExpr expr, DataType targetType) {
+    requireNonNull(expr, "aggregation expr is null");
+    aggregates.add(Pair.create(expr, targetType));
+    return this;
+  }
+
+  public List<TiExpr> getAggregates() {
+    return aggregates.stream().map(p -> p.first).collect(Collectors.toList());
+  }
+
+  public List<Pair<TiExpr, DataType>> getAggregatePairs() {
+    return aggregates;
+  }
+
+  /**
+   * add a order by clause to select query.
+   *
+   * @param byItem is a TiByItem.
+   * @return a SelectBuilder
+   */
+  TiDAGRequest addOrderByItem(TiByItem byItem) {
+    orderByItems.add(requireNonNull(byItem, "byItem is null"));
+    return this;
+  }
+
+  List<TiByItem> getOrderByItems() {
+    return orderByItems;
+  }
+
+  /**
+   * add a group by clause to select query
+   *
+   * @param byItem is a TiByItem
+   * @return a SelectBuilder
+   */
+  public TiDAGRequest addGroupByItem(TiByItem byItem) {
+    groupByItems.add(requireNonNull(byItem, "byItem is null"));
+    return this;
+  }
+
+  public List<TiByItem> getGroupByItems() {
+    return groupByItems;
+  }
+
+  /**
+   * Field is not support in TiDB yet, for here we simply allow TiColumnRef instead of TiExpr like
+   * in SelectRequest proto
+   * <p>
+   * <p>This interface allows duplicate columns and it's user's responsibility to do dedup since we
+   * need to ensure exact order and items preserved during decoding
+   *
+   * @param column is column referred during selectReq
+   */
+  public TiDAGRequest addField(TiColumnRef column) {
+    fields.add(requireNonNull(column, "columnRef is null"));
+    return this;
+  }
+
+  public List<TiColumnRef> getFields() {
+    return fields;
+  }
+
+  /**
+   * set key range of scan
+   *
+   * @param ranges key range of scan
+   */
+  public TiDAGRequest addRanges(List<Coprocessor.KeyRange> ranges) {
+    keyRanges.addAll(requireNonNull(ranges, "KeyRange is null"));
+    return this;
+  }
+
+  public void resetRanges(List<Coprocessor.KeyRange> ranges) {
```

Comment:
If keyRange is only used inside class. This is fine. If it is also used outside this class, this could be problematic. 
You want to new a array list instead of just clearall.

Reply:
`keyRanges` is final here so that we cannot new an `ArrayList` and assign it to `keyRanges`. Now that it's final, calling `clear()`'s behavior is the same both inside and outside of this class.

Reply:
As long as `keyRanges` is used inside this class, `clear` is fine.

Reply:
Ok